python/cudf/dataframe/dataframe.py

```diff
@@ -469,18 +469,22 @@ def take(self, positions, ignore_index=False):
             out[col] = self[col].take(positions, ignore_index=ignore_index)
         return out
 
-    def copy(self):
+    def copy(self, deep=True):
         """
         Returns a copy of this dataframe
         """
         df = DataFrame()
         df._index = self._index
         df._size = self._size
-        df._cols = self._cols.copy()
+        if deep:
+            for k in self._cols:
+                df[k] = self._cols[k].copy()
+        else:
+            df._cols = self._cols
```

Comment:
This does a pass by reference instead of pass by copy, so if a user assigns new columns to either dataframe it will populate both dataframes.

Reply:
We need to create a new OrderedDict object and populate it with references to the items in the original OrderedDict.

Reply:
The `else:` clause here is a shallow copy, isn't pass by reference what we want?

Reply:
The OrderedDict here is essentially the underlying structure of the DataFrame, if we pass this by reference assigning columns would affect both dataframes which isn't the expected behavior.